### PR TITLE
[DSV3] Apply TP on DSV3

### PR DESCRIPTION
**Pull request by wwwjn**
## Context
Mostly adapted from llama4, change the TP plan based on the difference between deepseek-v3 and llama.

Thanks @tianyu-l for the detailed walk through about deepseek-v3 attention model and TP plan! This diff is currently based on #1324 , and we want to extract the MoE model in DSV3 and llama4 in a shared place.

Now we have:
1. FSDP
2. Activation Checkpointing
3. TP
4. CP in progress (hang due to some reason)

## Next Step:
1. Make CP work

## Verification
There are minor issue with the numerical verification: With deterministic seed, the loss is not identical. I used `AdamW` optimizer.

1. FSDP degree=4 (blue line)
2. FSDP degree=4, TP degree = 2 (orange line)

Screenshot 2025-07-01 at 5 38 50 PM

With `Adam` optimizer, the loss is **exactly the same**:
Screenshot 2025-07-02 at 1 26 32 PM

### CODE REVIEW

**Review by tianyu-l — APPROVED** [review]
Could you verify numerical correctness by comparing e.g. FSDP 2 vs. FSDP 2 + TP 2, using the same seed checkpoint?

File: torchtitan/models/deepseek_v3/model/model.py [hunk 1]
```diff
@@ -278,12 +282,13 @@ def __init__(self, layer_id: int, model_args: DeepSeekV3ModelArgs):
...
         self.attention = Attention(model_args)
         self.attention_norm = nn.RMSNorm(model_args.dim, eps=model_args.norm_eps)
-        self.moe_norm = nn.RMSNorm(model_args.dim, eps=model_args.norm_eps)
-        self.moe = (
-            FeedForward(model_args.dim, model_args.inter_dim)
-            if layer_id < model_args.n_dense_layers
-            else MoE(model_args)
-        )
+        self.ffn_norm = nn.RMSNorm(model_args.dim, eps=model_args.norm_eps)
+        self.moe_enabled = layer_id < model_args.n_dense_layers
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/model/model.py]: This line is wrong , should be opposite :(

File: torchtitan/models/deepseek_v3/model/model.py [hunk 2]
```diff
@@ -311,7 +320,7 @@ def __init__(self, model_args: DeepSeekV3ModelArgs):
         self.max_seq_len = model_args.max_seq_len
         self.tok_embeddings = nn.Embedding(model_args.vocab_size, model_args.dim)
         self.register_buffer(
-            "freqs_cis", precompute_freqs_cis(model_args), persistent=False
+            "freqs_cis", precompute_freqs_cis(model_args), persistent=True
```
Comment (fegin) [inline torchtitan/models/deepseek_v3/model/model.py:322]: Why do we change this field to True? I know there is a bug that can cause issues when torch.compiler is used with non-persistent buffers.. Just curious why in this PR.

File: torchtitan/models/deepseek_v3/model/moe.py [hunk 3]
```diff
@@ -171,9 +211,12 @@ def forward(
         top_scores = (
             top_scores * self.route_sclaing_factor
         )  # must multiply the scaling factor
-
+        print("In TokenChoiceTopKRouter, top_scores shape: ", top_scores)
```
Comment (fegin) [inline torchtitan/models/deepseek_v3/model/moe.py]: nit: logger.info?

File: torchtitan/models/deepseek_v3/model/model.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/models/deepseek_v3/model/model.py:322]: Here I followed the comments in llama4: https://github.com/pytorch/torchtitan/blob/6b11290c1e45221e097aa4a54d4ab52bf067d0b5/torchtitan/experiments/llama4/model/model.py#L382-L389

File: torchtitan/models/deepseek_v3/model/moe.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/models/deepseek_v3/model/moe.py]: This should be removed , thanks for pointing out!

**Comment by wwwjn** [comment]
Note for reviewer: Because this diff is rebased onto #1324, there are bunch of files are being changed. Please focus on changes under `torchtitan/model/deepseek_v3/`, and all the changes needed for TP is under this folder.
